### PR TITLE
[notifications][android] prevent crashing in standalone apps when receiving interactive notifications

### LINKED ISSUE #10532: [Notifications] Interactive expo-notifications (local) seems crash standalone Android app
### output of `expo diagnostics` 
  Expo CLI 3.27.12 environment info:
    System:
      OS: macOS 10.15.6
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 12.18.3 - /usr/local/bin/node
      Yarn: 1.22.4 - /usr/local/bin/yarn
      npm: 6.14.8 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.9.3 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.0, DriverKit 19.0, macOS 10.15, tvOS 14.0, watchOS 7.0
      Android SDK:
        API Levels: 23, 24, 25, 26, 27, 28, 29
        Build Tools: 25.0.3, 27.0.3, 28.0.3
        System Images: android-28 | Google APIs Intel x86 Atom, android-28 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: 3.6 AI-192.7142.36.36.6241897
      Xcode: 12.0.1/12A7300 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^39.0.0 => 39.0.3 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: https://github.com/expo/react-native/archive/sdk-39.0.0.tar.gz => 0.63.2 
      react-native-web: ~0.13.7 => 0.13.13 
    npmGlobalPackages:
      expo-cli: 3.27.12
    Expo Workflow: managed

The platform(s) targeting:
Android and iOS, but only Android standalone app crashes. iOS seems working just fine on expo app and on standalone app.  The testing devices I use are two low-end Honor phones.

### Reproducible Demo

https://snack.expo.io/@takwwong5/notificationtesting

You may init a new expo, copy the code to it, npm install all packages and run it.  You will find that all interactive notifications should run fine on both android and ios, but if you build it and run it on standalone android app. it crashes when a local notification is about to be triggered.  I used android studio to look for a log.  It just said Notification something can not be cast to something (I cannot remember now sorry).

### PR DESCRIPTION
**Pull request by cruzach**
# Why

(It looks like scoping changes in https://github.com/expo/expo/pull/9287 may have affected notifications in standalone apps)

closes https://github.com/expo/expo/issues/10532

The crash was caused by `ScopedCategoryAwareNotificationBuilder.java` trying to cast a NotificationRequest to a `ScopedNotificationRequest`.  We guard against that now with a check, but after fixing that a few other issues came up:

- ~remote notifications in **standalone** apps were coming through as "scoped" bc `ExpoFcmMessagingService.java` would always return a `ScopedNotificationRequest`. To address that, I added a `ScopedExpoFcmMessagingService` that we rely on in the client, and then in XDL during android shell app modifications, we swap that value for `ExpoFcmMessagingService` in standalone apps.~ Edit: moved this change to its own PR - https://github.com/expo/expo/pull/10625

- Since `ScopedCategoryAwareNotificationBuilder` subclasses `CategoryAwareNotificationBuilder`, it was erroneously adding actions twice (since it would call `super.createBuilder`). Simple fix was to extract that action-adding logic of `CategoryAwareNotificationBuilder.createBuilder` out to an `addActionsToBuilder` method, and override that in `ScopedCategoryAwareNotificationBuilder` instead.

# How

~First commit adds `ScopedExpoFcmMessaging` for the client, and reserves `ExpoFcmMessagingService` for standalones ([XDL change here](https://github.com/expo/expo-cli/pull/2772)). this addresses the first bullet listed above~ ~ Edit: moved this change to its own PR - https://github.com/expo/expo/pull/10625

 (second commit is just import optimizing)

Third commit makes the changes explained in the second bullet above

# Test Plan

Tested in standalone android shell app and bare app, no crash

### CODE REVIEW

**Review by sjchmiela — COMMENTED** [review]
Regarding

> Since `ScopedCategoryAwareNotificationBuilder` subclasses `CategoryAwareNotificationBuilder`, it was erroneously adding actions twice (since it would call `super.createBuilder`).

how come it would add actions twice if `CategoryAwareNotificationBuilder` would look for unscoped `categoryId` (for which I guess there would be no results) and `ScopedCategoryAwareNotificationBuilder` would look for scoped `categoryId`?

**Comment by cruzach** [comment]
>how come it would add actions twice if CategoryAwareNotificationBuilder would look for unscoped categoryId (for which I guess there would be no results) and ScopedCategoryAwareNotificationBuilder would look for scoped categoryId?

sorry, I should've specified- this is only for standalone apps. In this case, `categoryId` is equal to the scoped category ID, but this only happens if a standalone app _does_ get a `ScopedNotificationRequest` (which is possible, hence https://github.com/expo/expo/pull/10625). If we address the issue where standalone apps can get a `ScopedNotificationRequest`, then we can instead just check `if (requester instanceof ScopedNotificationRequest) { do the category things }` in `ScopedCategoryAwareNotificationBuilder`

File: android/expoview/src/main/java/host/exp/exponent/notifications/ScopedCategoryAwareNotificationBuilder.java [hunk 1]
```diff
@@ -26,33 +26,33 @@ public ScopedCategoryAwareNotificationBuilder(Context context, @NonNull SharedPr
...
+        builder.addAction(super.buildTextInputAction((TextInputNotificationAction) action));
+      } else {
+        builder.addAction(super.buildButtonAction(action));
       }
     }
-
-    return builder;
   }
-}
+}
```
Comment (sjchmiela) [inline android/expoview/src/main/java/host/exp/exponent/notifications/ScopedCategoryAwareNotificationBuilder.java:58]:
```suggestion
}

```

**Review by sjchmiela — COMMENTED** [review]
I'm still not sure I get the scenario where actions would be added twice, but I don't think this matters that much, unwrapping `categoryId` in the builder makes sense! 👍 

Plus comment about using `super.addActionsToBuilder`

File: android/expoview/src/main/java/host/exp/exponent/notifications/ScopedCategoryAwareNotificationBuilder.java [hunk 2]
```diff
@@ -26,33 +26,33 @@ public ScopedCategoryAwareNotificationBuilder(Context context, @NonNull SharedPr
...
+      Log.e("expo-notifications", String.format("Could not read category with identifier: %s. %s", scopedCategoryIdentifier, e.getMessage()));
+      e.printStackTrace();
+    }
+    for (NotificationAction action : actions) {
+      if (action instanceof TextInputNotificationAction) {
+        builder.addAction(super.buildTextInputAction((TextInputNotificationAction) action));
+      } else {
+        builder.addAction(super.buildButtonAction(action));
       }
     }
```
[the remark this answers is not in the document]
Comment (cruzach) [reply android/expoview/src/main/java/host/exp/exponent/notifications/ScopedCategoryAwareNotificationBuilder.java]: yep 👍

**Pull request merged**